src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -189,13 +189,45 @@ private NamedTypeSymbol ConstructAnonymousDelegateImplementationSymbol(Anonymous
             // If all parameter types and return type are valid type arguments, construct
             // the delegate type from a generic template. Otherwise, use a non-generic template.
             bool useUpdatedEscapeRules = Compilation.SourceModule.UseUpdatedEscapeRules;
-            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr))
+            if (allValidTypeArguments(useUpdatedEscapeRules, typeDescr, out var needsIndexedName))
             {
                 var fields = typeDescr.Fields;
                 Debug.Assert(fields.All(f => hasDefaultScope(useUpdatedEscapeRules, f)));
 
-                bool returnsVoid = fields.Last().Type.IsVoidType();
+                bool returnsVoid = fields[^1].Type.IsVoidType();
                 int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
+                for (int i = 0; i < nTypeArguments; i++)
+                {
+                    typeArgumentsBuilder.Add(fields[i].TypeWithAnnotations);
+                }
+                var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
+
+                // Delegate type cannot be fully serialized to its name (e.g., it has default parameter values).
+                if (needsIndexedName)
+                {
+                    // Replace field types (where type parameters will be) by placeholders.
+                    var genericFieldTypeBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(fields.Length);
+                    for (var i = 0; i < nTypeArguments; i++)
+                    {
+                        genericFieldTypeBuilder.Add(default);
+                    }
```

Comment:
>} [](http://example.com/codeflow?start=20&length=1)

Consider separating blocks from other statements with empty line. For the sake of readability

Reply:
I was thinking about this as one block creating `genericTypeDescr`. Added some spaces if that helps.